scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -607,9 +611,14 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
 
             if (needOutdent)
               (sepRegions.tail, mkOutdent(lastNewlinePos))
-            else if (needIndent)
-              (RegionIndent(nextIndent, prev.is[KwMatch]) :: sepRegions, mkIndent(lastNewlinePos))
-            else if (canProduceLF) {
+            else if (needIndent) {
+              val nextRegion =
+                if (sepRegions.headOption.exists(_.isEnum))
+                  RegionIndentEnum(nextIndent)
```

Comment:
Are we doubling the region in case of `=> <indent>` ? We should not be inserting indentation at all after => if there is already an enum region.

Reply:
We put `Indent` for any template in this case. For `class A{ self => ..` and others.
It seems that it will be harder no to insert it after `SelfType =>` than preserving the enum region type in this case.

Reply:
`class A{ self => ..` <- this one for sure
However, maybe it doesn't make sense here:
```scala
enum A:
  self => ..
```
?

Reply:
Ha, you right!

I found the case that doesn't work. 
```scala
enum A:
  self =>
    case B, C
def foo: Int = ???
//^ error: outdent expected but def found
```

Moreover, Dotty parser has exactly the same hack as you propose: https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/parsing/Parsers.scala#L3836


Probably will port this trick.

Reply:
Reworked this by adding `undoIndent` method to `TokenIterator`